### PR TITLE
fossid-webapp: Introduce pattern options for naming projects and scans

### PR DESCRIPTION
**Pull request by nnobelis**
These patterns allow to control the name when a scan or project is created.
When a pattern is omitted, a default naming convention is used.

Signed-off-by: Nicolas Nobelis <nicolas.nobelis@bosch.io>

### CODE REVIEW

File: model/src/main/resources/reference.conf [hunk 1]
```diff
@@ -81,6 +81,22 @@ ort {
...
       }
+
+      FossId {
+        serverUrl = "https://fossid.example.com/instance/"
+        user = "user"
+        apiKey = "XYZ"
+
+        namingProjectPattern = "$Var1_$Var2"
+        namingScanPattern = "$Var1_#projectBaseCode_$Var3"
+        namingVar1 = "myOrg"
```
Comment (oheger-bosch) [inline model/src/main/resources/reference.conf:92]: Is this example correct? I would have expected that the variable must be named `namingVariable1`.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 2]
```diff
@@ -282,7 +309,8 @@ class FossId(
 
                 val url = pkg.vcsProcessed.url
                 val revision = pkg.vcsProcessed.revision.ifEmpty { "HEAD" }
-                val projectCode = convertGitUrlToProjectName(url)
+                val baseProjectCode = convertGitUrlToProjectName(url)
```
Comment (oheger-bosch) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:312]: I find these variable names confusing. What is now called `projectCode`, isn't this actually the `projectName`? And `projectBaseCode` could then be `projectCode` (or `projectUrl`, `projectId` ?).

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 3]
```diff
@@ -296,21 +324,24 @@ class FossId(
                 checkNotNull(scans)
 
                 val existingScan = scans.sortedByDescending { it.id }.find { scan ->
-                    scan.gitBranch == revision && scan.gitRepoUrl == url
+                    // The scans in the server contain the url with the credentials so we have to remove it for the
```
Comment (oheger-bosch) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:327]: This change does not seem to be related to the naming conventions, or is it?

File: scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * This class provides names for projects and scans when the FossID scanner creates them, following a given pattern.
+ * If one or both patterns is null, a default naming convention is used.
+ *
+ * [namingScanPattern] and [namingProjectPattern] are patterns describing the name using variables, e.g. "$var1_$var2".
+ * Variable values are given in the map [namingConventionVariables].
+ *
+ * There also are builtin variables. Builtin variables are are prefixed in the pattern with "#" e.g. "$var1_#builtin".
+ * Available builtin variables:
+ * * **projectBaseCode**: The name of the project (i.e. the part of the URL before .git).
```
Comment (oheger-bosch) [inline scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt:36]: Would it make sense to have the current timestamp as a built-in variable?

File: scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+    /**
+     * Replace the naming convention variables with their values. Used for projects and scans.
+     */
+    private fun replaceNamingConventionVariables(
+        namingConventionPattern: String, baseProjectCode: String,
+        namingConventionVariables: Map<String, String>
+    ): String {
+        log.info { "Parametrizing the name with naming=$namingConventionPattern, baseProjectCode=$baseProjectCode" }
+        var replacedName = namingConventionPattern
```
Comment (oheger-bosch) [inline scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt:62]: I guess, this can be simplified. You can create a map that contains both the built-in and the user-defined variables. Then the replacement can be done in a single step. You can also use `fold`, e.g.:
```
val replacedName = namingConventionVariables.fold(namingConventionPattern) { str, variable ->
  str.replace("\$${variable.key}", variable.value)
}
```

File: model/src/main/resources/reference.conf (hunk 1 above)
Comment (nnobelis) [reply model/src/main/resources/reference.conf:92]: No, all variables are prefixed with `naming` but the prefix is not in the variable name.
I did so to be able to group everything naming convention-related in the `ort.conf`, since it's no possible to have structured data such as:
```
        namingConvention {
          projectPattern = "$Var1_$Var2"
          scanPattern = "$Var1_#projectBaseCode_$Var3"
          variables {
            Var = "myOrg"
          }
        }
```

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 2 above)
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt]: Yes you are right, thinks got messy when I tried to have FossID semantic (scanCode, projectCode) and the concept of "naming convention".
I revamped it and would appreciate your opinion.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 6]
```diff
@@ -186,6 +196,23 @@ class FossId(
...
+            .filterKeys { it.startsWith(NAMING_CONVENTION_VARIABLE_PREFIX) }
+            .mapKeys { it.key.substringAfter(NAMING_CONVENTION_VARIABLE_PREFIX) }
+
+        namingProjectPattern?.let {
+            log.info { "Naming pattern for projects is $namingProjectPattern." }
+        }
+
+        namingScanPattern?.let {
+            log.info { "Naming pattern for scans is $namingScanPattern." }
+        }
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:212]: Nit: Maybe move these straight to the variable assignment above for more compact code?

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 2 above)
Comment (oheger-bosch) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt]: Yes, the new names are better. Only the `createScan()` function still uses the old ones as parameters names.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 7]
```diff
@@ -360,9 +391,11 @@ class FossId(
     /**
      * Create a new scan in the FossID server and return the scan code.
      */
-    private suspend fun createScan(projectCode: String, url: String, revision: String): String {
-        val formatter = DateTimeFormatter.ofPattern("'$projectCode'_yyyyMMdd_HHmmss")
-        val scanCode = formatter.format(LocalDateTime.now())
+    private suspend fun createScan(projectCode: String, baseProjectCode: String, url: String, revision: String):
+            String {
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:395]: Nit: Please rather format like

    private suspend fun createScan(
        projectCode: String, baseProjectCode: String, url: String, revision: String
    ): String {

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 8]
```diff
@@ -186,6 +196,23 @@ class FossId(
 
         log.info { "waitForResult parameter is set to '$waitForResult'" }
 
+        val namingProjectPattern = fossIdScannerOptions["namingProjectPattern"]
+        val namingScanPattern = fossIdScannerOptions["namingScanPattern"]
+
+        val namingConventionVariables = fossIdScannerOptions
+            .filterKeys { it.startsWith(NAMING_CONVENTION_VARIABLE_PREFIX) }
+            .mapKeys { it.key.substringAfter(NAMING_CONVENTION_VARIABLE_PREFIX) }
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:204]: From just reading the code, this looks strange: You're basically just stripping prefixed from key names. Which raises the question: Why bother adding these prefixes in the first place if they get stripped anyway?

Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:208]: Please see my answer above to Oliver:

> I did so to be able to group everything naming convention-related in the ort.conf, since it's no possible to have structured data

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 3 above)
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:323]: Moved to another commit.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 9]
```diff
@@ -320,7 +320,10 @@ class FossId(
                 checkNotNull(scans)
 
                 val existingScan = scans.sortedByDescending { it.id }.find { scan ->
-                    scan.gitBranch == revision && scan.gitRepoUrl == url
+                    // The scans in the server contain the url with the credentials so we have to remove it for the
+                    // comparison. If we don't, the scans won't be matched if the password changes!
+                    val urlWithoutCredentials = scan.gitRepoUrl!!.substringAfter("@")
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:325]: This also strips the protocol part, so does the comparison really work? Can't / shouldn't we make use of `replaceCredentialsInUri()` instead?

File: scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt (hunk 5 above)
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt:62]: 1. the built-ins are prefix with "#" so you cannot replace variables and build-ins in one step...
2. I don't find `.fold` for `Map`. Do you mean `foldTo` for `MutableMap` ?

File: scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt (hunk 4 above)
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt]: YAGNI but ok :)

Comment (oheger-bosch) [reply scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt]: I think it is already used for the default pattern; so if you just slightly wanted to adapt the default pattern, you would need such a variable.

File: scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt (hunk 5 above)
Comment (oheger-bosch) [reply scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt:66]: at 1: This depends how you organize the keys in the map and the replacement logic, but I won't insist.
at 2: You can use `Map.entries().fold()`.

File: model/src/main/resources/reference.conf (hunk 1 above)
Comment (oheger-bosch) [reply model/src/main/resources/reference.conf:92]: Ok, got this; but from my understanding, the constant `NAMING_CONVENTION_VARIABLE_PREFIX` has the value "namingVariable", and you filter for keys starting with this value. So the key `namingVar1` used in the example would not be matched - or am I wrong?

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 10]
```diff
@@ -186,6 +196,19 @@ class FossId(
 
         log.info { "waitForResult parameter is set to '$waitForResult'" }
 
+        val namingProjectPattern = fossIdScannerOptions["namingProjectPattern"]?.apply {
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:199]: Nit: Using `let` (and `it` in the next line) would be a bit more common for this use-case, IMO. Same below.

Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:199]: Understood but then you have to return something from `let` i.e.

```
val namingProjectPattern = fossIdScannerOptions["namingProjectPattern"]?.let {
            FossId.log.info { "Naming pattern for projects is $it." }
            it
}
```
which is more code or am I missing something ?

File: scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt (hunk 5 above)
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt]: It's ok...I reworked the code to do it in one operation.

Thanks for the suggestions !

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 10 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:199]: My bad, you're right. I missed that.

File: model/src/main/resources/reference.conf (hunk 1 above)
Comment (nnobelis) [reply model/src/main/resources/reference.conf]: You are of course right. Fixed !

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 11]
```diff
@@ -320,7 +320,10 @@ class FossId(
                 checkNotNull(scans)
 
                 val existingScan = scans.sortedByDescending { it.id }.find { scan ->
-                    scan.gitBranch == revision && scan.gitRepoUrl == url
+                    // The scans in the server contain the url with the credentials so we have to remove it for the
+                    // comparison. If we don't, the scans won't be matched if the password changes!
+                    val urlWithoutCredentials = scan.gitRepoUrl?.replaceCredentialsInUri()
+                    scan.gitBranch == revision && urlWithoutCredentials == url
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:326]: One last commit message nit: "GIT" should be "Git" (as it's not an acronym).

**Review by sschuberth — APPROVED** [review]

**Comment by sschuberth** [comment]
Merging despite the unrelated `PipFunTest` failure which got fixed by https://github.com/oss-review-toolkit/ort/pull/4262.
